src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -14,592 +14,549 @@ namespace Microsoft.CodeAnalysis.CSharp
+            // we go, pulling them all into a string literal.
+            CollectAndVisitConcatArguments(originalOperator, visitedCompoundAssignmentLeftRead: null, out var arguments);
 
-            FlattenConcatArg(loweredLeft, leftFlattened);
-            FlattenConcatArg(loweredRight, rightFlattened);
+            return RewriteStringConcat(originalOperator.Syntax, arguments);
+        }
 
-            if (leftFlattened.Any() && rightFlattened.Any())
+        private BoundExpression RewriteStringConcat(SyntaxNode originalSyntax, ArrayBuilder<BoundExpression> visitedArguments)
+        {
+            // The strategy of this rewrite is to pull out all the individual arguments of the concat, visit them as appropriate,
+            // and then construct the final concat all in one go. There are a few different lowering patterns that we take:
+            //
+            // 1. If we're in an expression tree, we have to follow strict binary operator rules. Nothing fancy to do, just do a recursive
+            //    rewrite into nested `string.Concat` calls.
+            // 2. If all the added expressions are constants, we can fold them all together and return a single constant.
+            // 3. If all the added expressions are strings, then we want to use one of the `string.Concat(string)`-based overloads: if 4 or less,
+            //    we'll use one of the hardcoded overloads. Otherwise, we'll use `string.Concat(string[])`.
+            // 4. If all the added expressions are strings or chars, we can use the `string.Concat(ReadOnlySpan<char>)`-based overloads. If there are
+            //    more than 4 arguments, we will instead fall back to `string.Concat(string[])`.
+            // 5. If all the added expressions are strings or objects, we'll use the `string.Concat(string)`-based overloads, and call ToString on the
+            //    arguments to avoid boxing structs by converthing them into objects. If there are more than 3, we'll use `string.Concat(string[])`.
+
+            if (visitedArguments.Count == 0)
             {
-                folded = TryFoldTwoConcatOperands(leftFlattened.Last(), rightFlattened.First());
-                if (folded != null)
-                {
-                    rightFlattened[0] = folded;
-                    leftFlattened.RemoveLast();
-                }
+                // All the arguments were null or the empty string. We can just return a constant empty string.
+                visitedArguments.Free();
+                return _factory.StringLiteral(string.Empty);
+            }
+            else if (visitedArguments is [{ ConstantValueOpt: { IsString: true } or { IsChar: true } } arg])
+            {
+                // We were able to fold a constant, so we can just return that constant.
+                visitedArguments.Free();
+                return arg as BoundLiteral ?? VisitExpression(arg);
```

Comment:
>arg as BoundLiteral [](http://example.com/codeflow?start=23&length=19)

It looks like we might end up with a char literal here. Shouldn't we make a string out of it?

Reply:
Good catch, adding `ConstantCharPlusNull` to cover this.